Fix: Windows long filename error

sepehr-rs opened the pull request:
Fixes #13346

This is my first attempt at addressing this issue. I’d appreciate any feedback or suggestions on how to improve the approach or make it more concise.

notatallshaw commented:
Hi @sepehr-rs, thanks for your PR, please be aware it may take some time for pip maintainers to review or respond to this PR as all maintainers are currently volunteers.

Though please let me know if you need any assistance fixing the pre-commit errors.

ichard26 commented on src/pip/_internal/commands/install.py:
```diff
@@ -775,9 +776,23 @@ def create_os_error_message(
         )
         parts.append(".\n")
 
+    # Windows raises EINVAL when a file or folder name exceeds 255 characters,
+    # even if long path support is enabled. Add a hint for that case.
```
```suggestion
    # Windows raises EINVAL when a file*name* or path segment exceeds 255
    # characters, even if long path support is enabled.
```

It wasn't immediately obvious to me what the difference between this check and the pre-existing hint was. That's my bad, but we can make this clearer :)

ichard26 commented on news/13346.bugfix.rst:
```diff
@@ -0,0 +1 @@
+This change will improve error hint for long filenames on Windows (EINVAL)
```
```
Provide a hint if a system error is raised involving long filenames or path segments on Windows. 
```

The changelog entry should be worded for the user. No need for "this change will [...]" as that implied.

ichard26 commented on src/pip/_internal/commands/install.py:
```diff
@@ -775,9 +776,23 @@ def create_os_error_message(
+        WINDOWS
+        and error.errno == errno.EINVAL
+        and error.filename
+        and any(len(part) > 255 for part in Path(error.filename).parts)
```
Nice use of pathlib!

sepehr-rs replied:
Thanks, glad to hear that :)!

ichard26 reviewed (changes requested):
Hello,

This looks generally right, but I have suggestions to tweak the changelog and comments. I also ask that you add a unit test to verify that this works properly. This function is tested here: https://github.com/pypa/pip/blob/e22047b60ea94eab4d8fafdbd0c7ced0e6372c79/tests/unit/test_command_install.py#L125

You simply need to add another case for this specific error.

Thanks for your interest in improving pip! I look forward to merging your contribution once it's ready.

sepehr-rs commented:
Hi @ichard26, sorry for the delay!

Thank you so much for your thoughtful suggestions and the time you took to review them. I've addressed all the requested changes and also added a platform check to skip the relevant test cases when not running on Windows. Since the issue is Windows-specific, I opted for this approach to keep things scoped and avoid duplicating test cases.
Looking forward to your feedback :)!

pfmoore reviewed (approved):
This looks OK to me. I confirmed that EINVAL is reported if you try to create a file with a part more than 255 characters long, so the check seems good to me.

ichard26 commented:
Sorry, I wasn't clear enough. I think it's still worth keeping the error messages for both scenarios separate. If any segment of the path exceeds the 255 character limit, then it's simply not going to work and enabling Long Paths won't help. This should be reflected in the error message. The error message from the previous version of this PR did. However, if the path overall exceeds 260 characters, then it's possible that Long Paths aren't enabled but should be. That possibility should be included in the error message. A good error message offers a hint for what to do, but it's also important to not lead the user down a rabbit hole of trying a solution _we know isn't going to help._

In other words, when I made that suggestion, I simply wanted you to tweak the `exc.errno == errno.EINVAL` condition to `exc.errno in (errno.EINVAL, errno.ENOENT)` in your **new** error. Literally just a one line change. 

Does that make sense? Sorry again, I don't mean to make you do a bunch of work.

sepehr-rs commented:
> Sorry, I wasn't clear enough. I think it's still worth keeping the error messages for both scenarios separate. If any segment of the path exceeds the 255 character limit, then it's simply not going to work and enabling Long Paths won't help. This should be reflected in the error message. The error message from the previous version of this PR did. However, if the path overall exceeds 260 characters, then it's possible that Long Paths aren't enabled but should be. That possibility should be included in the error message. A good error message offers a hint for what to do, but it's also important to not lead the user down a rabbit hole of trying a solution _we know isn't going to help._
> 
> In other words, when I made that suggestion, I simply wanted you to tweak the `exc.errno == errno.EINVAL` condition to `exc.errno in (errno.EINVAL, errno.ENOENT)` in your **new** error. Literally just a one line change.
> 
> Does that make sense? Sorry again, I don't mean to make you do a bunch of work. I'm just trying to guide/coach you through the reasoning process since you did ask for feedback.

Hi @ichard26, that makes total sense, thank you so much for taking the time to explain the distinction so clearly, and no worries at all! I really appreciate the guidance, it helps me learn a lot through this process.

I’ll go ahead and make that change to handle both `EINVAL` and `ENOENT` while preserving the distinct error messages as you suggested. I’ll push an update shortly. Thanks again for the review and support! :)
